DQM/SiPixelPhase1Heterogeneous/plugins/SiPixelPhase1CompareRecHitsSoA.cc

```diff
@@ -0,0 +1,174 @@
+
+class SiPixelPhase1CompareRecHitsSoA : public DQMEDAnalyzer {
+public:
+  explicit SiPixelPhase1CompareRecHitsSoA(const edm::ParameterSet&);
+  ~SiPixelPhase1CompareRecHitsSoA() override = default;
+  void dqmBeginRun(const edm::Run&, const edm::EventSetup&) override;
+  void bookHistograms(DQMStore::IBooker& ibooker, edm::Run const& iRun, edm::EventSetup const& iSetup) override;
+  void analyze(const edm::Event& iEvent, const edm::EventSetup& iSetup) override;
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+private:
+  const edm::ESGetToken<TrackerGeometry, TrackerDigiGeometryRecord> geomToken_;
+  const edm::ESGetToken<TrackerTopology, TrackerTopologyRcd> topoToken_;
+  const edm::EDGetTokenT<TrackingRecHit2DCPU> tokenSoAHitsCPU_;
+  const edm::EDGetTokenT<TrackingRecHit2DCPU> tokenSoAHitsGPU_;
+  const std::string topFolderName_;
+  const TrackerGeometry* tkGeom_ = nullptr;
+  const TrackerTopology* tTopo_ = nullptr;
+  MonitorElement* hnHits_;
+  MonitorElement* hBchargeL_[4];
+  MonitorElement* hBsizexL_[4];
+  MonitorElement* hBsizeyL_[4];
+  MonitorElement* hFchargeD_[2][3];
+  MonitorElement* hFsizexD_[2][3];
+  MonitorElement* hFsizeyD_[2][3];
+};
+
+//
+// constructors
+//
+
+SiPixelPhase1CompareRecHitsSoA::SiPixelPhase1CompareRecHitsSoA(const edm::ParameterSet& iConfig)
+    : geomToken_(esConsumes<TrackerGeometry, TrackerDigiGeometryRecord, edm::Transition::BeginRun>()),
+      topoToken_(esConsumes<TrackerTopology, TrackerTopologyRcd, edm::Transition::BeginRun>()),
+      tokenSoAHitsCPU_(consumes<TrackingRecHit2DCPU>(iConfig.getParameter<edm::InputTag>("pixelHitsSrcCPU"))),
+      tokenSoAHitsGPU_(consumes<TrackingRecHit2DCPU>(iConfig.getParameter<edm::InputTag>("pixelHitsSrcGPU"))),
+      topFolderName_(iConfig.getParameter<std::string>("topFolderName")) {}
+//
+// Begin Run
+//
+void SiPixelPhase1CompareRecHitsSoA::dqmBeginRun(const edm::Run& iRun, const edm::EventSetup& iSetup) {
+  tkGeom_ = &iSetup.getData(geomToken_);
+  tTopo_ = &iSetup.getData(topoToken_);
+}
+
+//
+// -- Analyze
+//
+void SiPixelPhase1CompareRecHitsSoA::analyze(const edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  const auto& rhsoaHandleCPU = iEvent.getHandle(tokenSoAHitsCPU_);
+  const auto& rhsoaHandleGPU = iEvent.getHandle(tokenSoAHitsGPU_);
+  if (!rhsoaHandleCPU.isValid() || !rhsoaHandleGPU.isValid()) {
+    edm::LogWarning("SiPixelPhase1CompareRecHitsSoA") << "No RecHits SoA found on CPU or GPU\n returning!" << std::endl;
+    return;
+  }
+  auto const& rhsoaCPU = *rhsoaHandleCPU;
+  const TrackingRecHit2DSOAView* soa2dCPU = rhsoaCPU.view();
+  auto const& rhsoaGPU = *rhsoaHandleGPU;
+  const TrackingRecHit2DSOAView* soa2dGPU = rhsoaGPU.view();
+
+  uint32_t nHitsCPU = soa2dCPU->nHits();
+  uint32_t nHitsGPU = soa2dGPU->nHits();
+  hnHits_->Fill(nHitsCPU,nHitsGPU);
+  auto detIds = tkGeom_->detUnitIds();
+  for (uint32_t i = 0; i < nHitsCPU; i++) {
+    float minD=99.;
+    uint32_t matchedHit=9999999;
```

Comment:
could 999999 be like a named constant, e.g. `invalidHit`  ?

Reply:
even better: couldn't we use a less arbitrary number, such as std::numeric_limits<uint32_t>::max() (which could still be introduced as a named constant for the sake of code readability)